### ISSUE TITLE
How to encode / use additional fonts?

### ISSUE THREAD

**Issue by d4tocchini** [makepad/makepad.github.io#8]
Is there a particular config/command you are using for msdfgen?  

Has the included fontencoder source deviated from [Chlumsky/msdfgen](https://github.com/Chlumsky/msdfgen)? wondering if output from latest (v1.4) ought to work.

Playing around with HUD like UI, need me some Inconsolata :)

@makepaddev, if helpful, I can PR in a few more open-source fonts.

**Comment by makepaddev** [comment]
Hey, so msdfgen i run on a font, and then the only thing i tweak is the 'scale' factor untill the font is artifact free (larger usually something like 1.6 1.8)

**Comment by makepaddev** [comment]
I made a single file compile out of it (makefile) you should be able to just overwrite some files and recompile it

**Comment by makepaddev** [comment]
fontencoder.cpp is my own use of msdfgen C++ API which does all my font file generation

**Comment by d4tocchini** [comment]
Thanks! It works.  

for anyone curious how, as setup:

```
cd ./fontencoder
chmod +x ./Makefile
./Makefile
```

which should compile a `fontencoder` binary, which can be used like so:  

```
./fontencoder {{path/to/awesome-font.ttf}} ../libs/fonts/{{awesome-font}}.font {{scale}} 0 126
```

I had to use a scale above `2.0` for inconsolata before it rendered as expected.

Unicodes 0 to 126 covered all the bases, can likely remove unused chars.

Font render quality varies on retina vs non-retina screens.

**Comment by makepaddev** [comment]
Yeah non retina is extremely hard to do right, its actually amazing that it isnt complete garbage. I'm just kinda waiting out the non-retina to go away tbh.

**Comment by makepaddev** [comment]
All non-retina fonts really need pixel hinting and all that jazz. Don't have that on a gpu, probably never will.

**Comment by d4tocchini** [comment]
I can only imagine...  

After some tweaking, a little bit extra boldness, looks solid on non-retina.  It's amazing you got this far on pure GPU!

**Comment by makepaddev** [comment]
Well with the great help of the guy that made the msdflib, or before that glyphy. This is font-engine-version 4 for what i'm doing, so i've gotten quite decent at the fiddly bits. But yes besides the drawcall packing shader compiler its definitely one of the enabling cornerstones of why makepad works.

**Comment by makepaddev** [comment]
What is on my todo is to have some kind of nice nonretina-to-retina curve for that boldness. I can't seem to find a 'single' value that works, like if you pick one that looks good on lowdpi its too bold on hidpi. etc.

**Comment by makepaddev** [comment]
Btw to answer your first question, i don't really need PR of more fonts, im trying to keep the repo as tiny as possible for as long as possible, big binaries mess that up.

**Comment by d4tocchini** [comment]
after seeing the compiled font file size, makes sense.  

is there a flag set somewhere when retina or not?  could be a simple config to the styles files

**Comment by makepaddev** [comment]
Hi, so the thing is 'retina mode' isnt just a single value its actually the zoom-factor. (browser zoom works perfectly in makepad through this)

**Comment by makepaddev** [comment]
That way you can just zoom in/out of a UI however you please, its extremely useful for most UIs.

**Comment by makepaddev** [comment]
Btw the generated fonts do zip quite nicely, so generally you'd use transport compression, since that also is JS-cost free

**Comment by makepaddev** [comment]
Oh also, in AR/VR mode (what im working on now) this whole 'retina' factor will go away entirely since it will be 3D space, so i really wanna keep things like retina-mode-ness mostly a float

**Comment by d4tocchini** [comment]
👍  awesome, precisely where I was hoping you were taking all this 

how stable is the core UI/view API? expect major/minor differences with AR/VR mode?

**Comment by d4tocchini** [comment]
BTW, i stumbled across this [interesting abandoned side-proj thingy](https://github.com/evanw/sky), it's the only other pure webgl code editor attempt I've seen.  albeit a different beast/stack altogether, but worth noting.  the guys at Figma are doing interesting things with their gl collaborative UI design tool.  

Also, is there a more proper place for this kind of non-specific convo?

**Comment by WolfieWerewolf** [comment]
@d4tocchini I think Evan is a founder of Figma. "Sky" is an interesting prototype but the Skew language couldn't possible be considered for any kind of production development.  A "one developer" language is destined for failure. There are, however, some other very interesting well backed strongly typed languages (not TypeScript) that compile to JS and make working with WebGL much more pleasant.  I honestly don't know how @makepaddev does it... writing this much JS by hand would do my head in.  I like JS but I also like compile time checking and OO inheritance... Maybe I started with the "C"'s so I'm biased.  In any case @makepaddev your project has been an inspiration to me.  I keep regular tabs on your progress.  Thanks for sharing.
Kind Regards
/W

**Comment by makepaddev** [comment]
Hi so, figma is a hybrid html/GL UI, i hope they may change that someday. Otherwise its a very impressive project, although i don't agree with the compiled-from-C++ stack and other things, Evan is definitely extremely skilled. Makepad is an attempt to truly put the entire UI stack in GL and expose all the programmable bits 'easily' to the developer, as easy as writing CSS you write a little shader JS function that does what you need. I've pretty much completed the 'search' for how to do the UI kit, and the shader code, i hope. However the underlying fabric, the render API is going to get a big overhaul as i'm redesigning the shader compiler to be split into 2 parts. A user space part which creates a bytecode AST and a mainthread part that generates GLSL. This way i can platform-fix bugs in videocard drivers for particular problems, and otherwise bridge webGL1/2. One of the things i want the compiler to fill in is the absence of float rendertargets on most of the mobile space. So the renderAPI should give you a way to write computation shaders regardless of the existence of float targets. If i don't fix this centrally most of the people will be missing out on all the fun of gpgpu atleast when viewing fun little programs. So on the stability, i think the UI bit is going to be quite stable / minor, but one reason i put this warning in the readme is because i'm not done evolving major pieces yet.

**Comment by makepaddev** [comment]
W: thanks for your reply, i hope to one day become more useful with makepad than inspiration, but until then i'll settle for that :)

**Comment by WolfieWerewolf** [comment]
@makepaddev like you said you're still working on the underlying and you made that clear with your "no not use" caveat ;-).  I took you seriously.  I'm working on something similar but for a completely different use case.  I feel WebGL has been vastly under utilized in the 2D space especially for Ui Kits.  And the 3D guys don't seem to have much use for a proper hardware accelerated GUI for some reason.... When you get through the underlying I'll be very interested in taking a look at how you solved certain problems that I'm sure I'm also working on.
Again very nice work mate...

**Comment by makepaddev** [comment]
Yep i agree on the underutilised for 2D, trying to fix that part :) Yea let me know if you have something to show here or on twitter, happy to check it out or give feedback
